### PR TITLE
bot, irc: IRC backend using asyncio

### LINKED ISSUE #1414: Rewrite to use asyncio instead of asynchat
Python's `asyncore` and `asynchat` modules are deprecated as of 3.6. The `asyncio` module was added in 3.4, and Sopel should probably be rewritten on top of that when Python 2.x support is dropped.

I don't think there's a specific timeline for that happening yet. Maybe whenever Python releases a stable version that no longer supports `asyncore` and `asynchat`, it'll be a good time to make Sopel require Python 3.4+. That probably won't happen for years to come, but I think it's good to get this on the radar.

No milestone until Python actually removes the modules. The next major release of Sopel after that happens will include this rewrite. (Anyone want to place bets on what the version number will be? :joy_cat:)

### PR DESCRIPTION
**Pull request by Exirel**
### Description

**TL;DR**: Tin. Close #1414.

Wouhou! This is it, that's the big PR for Sopel 8 that replaces the backend made with the now deprecated `asynchat` (and `asyncore`) module by a backend made with `asyncio`—and it wasn't easy. It might be complicated to review this, even more with all the added stuffs that feel unnecessary (like a heavy dose of type annotation).

I've tested it myself on two networks: Libera (with SSL on), and a private network (without SSL), and it all worked fine, ping timeout was handled properly, server timeout as well, ctrl+c is working fine as far as I can tell, even tho the sequence is a bit complicated if you try to interrupt the bot at the startup sequence (servers tend to wait quite some time before replying to a `QUIT` command and closing the connection).

So what's in the box?

* Asynchat based backends have been removed, and replaced by the `AsyncioBackend`
* A lot of parameters have been moved from `AbstractBackend.run` to the `__init__` of `AsyncioBackend`
* It is now easier to know what are the connection parameters, all put in one single method
* SSL is now managed with an SSL Context (this makes parts of #2246 obsolete, sadly, and I haven't put the TLS minimum version (yet))
* `asyncio`'s high level API is used when possible, instead of working with the low-level API; this means that I didn't dive into Protocol, Transport, Socket, etc. and I decided to use the `StreamReader` and `StreamWriter` object
* There is still a bit of `asyncio.Task` and `asyncio.TimerHandle` management involved (mostly for thread-safety and for timeout management)
* Signal handlers are now managed by the backend itself (it was impossible otherwise)
* Methods required on the `AbstractBot` ABC class and `AbstractIRCBackend` have been created or moved accordingly
* I had zero trouble with exiting the bot from the `start` command, so I went maybe a bit too ballistic on that one—but given it now runs an `EventLoop`... who knows?

What's missing?

* TLS management wasn't ported (yet) from #2246 
* I'm  not very happy with how `on_irc_error` and `on_error` callbacks are used (when/how)
* I know it works, but I would like to be a bit more sure about the timeout management
* I've no idea how to deal with CNAME check; I think @half-duplex asked about it on #2246 and I've to say... why did we check that? I'm not sure it's a good thing to keep...
* I know it's possible but I haven't tried yet to run every single rule handlers into an asyncio task, trough [asyncio.to_thread](https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread), and making `bot.on_message` a coroutine. That would involve a *lot* of change, while being 100% backward compatible with the current plugins & rules system. The end goal? To allow plugin to define `async` plugin callables :sunglasses: 
* A better documentation.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

**Comment by Exirel** [comment]
About AsyncioBackend parameters:

image

**Review by dgw — COMMENTED** [review]
I actually made it through all the code, no thanks to Refined GitHub messing up my Split View (so I disabled the extension). Along the way I ran into a surprisingly low number of "WTF" moments (but don't worry, some did still happen).

My favorite thing about this might be that there's only one backend class now, instead of separate SSL/non-SSL classes. That means the bot doesn't have to dynamically choose a class to use at runtime. Apparently there was no compelling technical reason to build it that way (even if the forced polymorphism was neat just because you don't get to do things like that very often).

> ctrl+c is working fine as far as I can tell, even tho the sequence is a bit complicated if you try to interrupt the bot at the startup sequence (servers tend to wait quite some time before replying to a `QUIT` command and closing the connection)

I've seen "problems" like this with the current implementation too, so don't worry about it. You're maintaining feature parity. :P

> * I know it's possible but I haven't tried yet to run every single rule handlers into an asyncio task, trough [asyncio.to_thread](https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread), and making `bot.on_message` a coroutine. That would involve a _lot_ of change, while being 100% backward compatible with the current plugins & rules system. The end goal? To allow plugin to define `async` plugin callables 😎

This is a good future goal. And it _must_ be a _future_ goal until we are ready to drop support for Python \<=3.8, because `asyncio.to_thread()` isn't available until Python 3.9.

**Comment by Exirel** [comment]
> This is a good future goal. And it _must_ be a _future_ goal until we are ready to drop support for Python <=3.8, because `asyncio.to_thread()` isn't available until Python 3.9.

Hm... yes, you're right about `to_thread`. However, we can already work with thread using [loop.run_in_executor()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor) instead, I was just too caught up into "the future" that I forgot that the convenient/shortcut function only exists in 3.9. 😁

File: sopel/irc/abstract_backends.py [hunk 1]
```diff
@@ -17,39 +23,67 @@ class AbstractIRCBackend(abc.ABC):
...
+        except UnicodeDecodeError:
+            # not Unicode; let's try CP-1252
+            try:
+                data = str(line, encoding='cp1252')
+            except UnicodeDecodeError:
+                # Okay, let's try ISO 8859-1
+                try:
+                    data = str(line, encoding='iso8859-1')
+                except UnicodeDecodeError:
+                    raise RuntimeError('Unable to decode data from server.')
```
[the remark this answers is not in the document]
Comment (Exirel) [reply sopel/irc/abstract_backends.py:78]: Because it doesn't make a lot of sense to log that here anyway, given that the caller of `decode_line` should be the one responsible for that: it has the raw line in bytes and an exception to deal with.

However, I don't think I log anything when calling `decode_line` later on, so that should be considered.

File: sopel/irc/backends.py [hunk 2]
```diff
@@ -5,346 +5,330 @@
...
+    ):
+        super().__init__(bot)
+        # connection parameters
+        self._host: str = host
+        self._port: int = port
+        self._source_address: Optional[Tuple[str, int]] = source_address
+        self._use_ssl: bool = use_ssl
+        self._certfile: Optional[str] = None
+        self._keyfile: Optional[str] = None
+        self._verify_ssl: bool = True
```
[the remark this answers is not in the document]
Comment (Exirel) [reply sopel/irc/backends.py:83]: I'd still have this as `True` by default anyway, because if `use_ssl` is `True`, you want it `True`, if it's `False`, it doesn't matter.

File: sopel/irc/backends.py [hunk 3]
```diff
@@ -5,346 +5,330 @@
...
-    def collect_incoming_data(self, data):
-        """Try to make sense of incoming data as Unicode.
+        .. seealso::
 
-        :param bytes data: the incoming raw bytes
+            The :meth:`~.decode_line` method is used to decode the IRC line
+            from :class:`bytes` to :class:`str`.
 
-        The incoming line is discarded (and thus ignored) if guessing the text
-        encoding and decoding it fails.
```
[the remark this answers is not in the document]
Comment (Exirel) [reply sopel/irc/backends.py:197]: The approach is better, but the lack of try/catch on calling decode_line isn't good. Will fix that.

File: sopel/irc/backends.py [hunk 4]
```diff
@@ -5,346 +5,330 @@
...
+                    certfile=self._certfile,
+                    keyfile=self._keyfile,
+                )
+
+            if not self._verify_ssl:
+                # deactivate SSL verification for hostname & certificate
+                ssl_context.check_hostname = False
+                ssl_context.verify_mode = ssl.CERT_NONE
+            else:
+                # verify location if verify_mode is not CERT_NONE
```
[the remark this answers is not in the document]
Comment (Exirel) [reply sopel/irc/backends.py:276]: As said on IRC, it all comes to the default value for both `verify_mode` and `check_hostname` based on `purpose` being `SERVER_AUTH`. Moreover, if you have `check_hostname=True`, then you cannot have something different than `verify_mode` set to `CERT_REQUIRED`.

Also, there is only 3 possible values:

* `CERT_NONE`: must be manually set
* `CERT_REQUIRED`: set by default on `SERVER_AUTH`
* `CERT_OPTIONAL`: same as `CERT_REQUIRED` for client context (which is the case for the purpose of `SERVER_AUTH`); otherwise allow the client to refuse the certificate (in server context)

So it's a comment that implies that you know a bit about how `ssl.create_default_context(ssl.Purpose.SERVER_AUTH)` works.

File: sopel/logger.py [hunk 5]
```diff
@@ -106,6 +106,11 @@ def setup_logging(settings):
                 'propagate': False,
                 'handlers': ['exceptionfile'],
             },
+            # asyncio logging
+            'asyncio': {
+                'level': 'DEBUG',
+                'handlers': ['console'],
+            },
```
[the remark this answers is not in the document]
Comment (Exirel) [reply sopel/logger.py:113]: Oh yeah, I forgot to check if it outputs anything if asyncio isn't used in debug mode (either by an env var or with using the debug parameter).

File: sopel/irc/__init__.py [hunk 6]
```diff
@@ -30,83 +30,91 @@
...
-    def name(self):
+    def name(self) -> str:
         """Sopel's "real name", as used for WHOIS responses."""
         return self._name
 
     @property
-    def config(self):
+    def config(self) -> Config:
         """The :class:`sopel.config.Config` for the current Sopel instance."""
         # TODO: Deprecate config, replaced by settings
```
[the remark this answers is not in the document]
Comment (Exirel) [reply sopel/irc/__init__.py:104]: Oh! I remember that.

The issue is that having using the word `config` for a variable name make it more difficult to work with the import of `from sopel import config`, and I tend to make a distinction in code between "config" (i.e. the model, the concept, the specification of the configuration file) and "settings" (the actual values, the concrete object), as in `settings = config.Config(...)`.

Having `bot.settings` instead of `bot.config` makes it less confusing to me. And yes, btw, I'm still mad at `sopel.bot` and `sopel.trigger` for that exact same reason...

**Comment by dgw** [comment]
Then it depends on whether you want to write the code twice, or wait until you can use the shortcut and write it only once. :P

File: sopel/irc/abstract_backends.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dgw) [reply sopel/irc/abstract_backends.py:79]: Probably want to use something other than `RuntimeError`? I don't think I was clear about _what part_ of the approach I don't like, which is the lack of logging and the choice of exception type. I don't care if the caller is responsible for logging the failure, but I think something like `ValueError` would make more sense.

File: sopel/irc/backends.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dgw) [reply sopel/irc/backends.py:276]: I could make a comment here about following the Zen of Python and setting explicit values even if they're identical to the defaults… But really, it's just the wording of the comment itself that bugs me, because I read about those defaults recently while looking at the other TLS PR and the comment still didn't make sense.

Maybe reordering the if/else conditions would make more sense: If verify_ssl, set that up; else disable verification in the SSL context. Would you be OK with that?

No suggestion, sorry. I'm on my phone, waiting for AAA to rescue our car that won't start. 🙃

File: sopel/irc/backends.py [hunk 7]
```diff
@@ -5,346 +5,330 @@
...
+                ssl_context.load_verify_locations(self._ca_certs)
```
Comment (half-duplex) [inline sopel/irc/backends.py:277]: This will raise `TypeError: cafile, capath and cadata cannot be all omitted` if `self._ca_certs=None`, so I don't think `Optional[]` is appropriate for `__init__.ca_certs` unless there's a check or the default is loaded here instead of core_section.py

Aside: Should we really be doing [`_find_certs()`](https://github.com/sopel-irc/sopel/blob/4322aa45d345327f27e1616af460937fa3e5e991/sopel/config/core_section.py#L26) instead of trusting Python to have a sane default set by upstream or distro packagers? This smells bad.

File: sopel/irc/backends.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (half-duplex) [reply sopel/irc/backends.py:276]: I think it's clear enough what's happening without the comment, or even better with the conditional flipped.

**Review by half-duplex — CHANGES_REQUESTED** [review]
I'm in favor of abandoning the CNAME stuff. I don't like mucking with things like hostname verification, and it was introduced specifically for Twitch, where it's easily worked around by updating configs to point where the CNAME points.

ℹ️ This drops `do_handshake_on_connect=False` and `suppress_ragged_eofs=True`, introduced in 614abe1a2b074a6869fd588e08427fc4480044b5. asyncio should handle the first, and the second is the [default](https://docs.python.org/3/library/ssl.html#ssl.wrap_socket).

I haven't played with this, since I need the other TLS changes from my PR, which I don't want to redo until this one is merged.

File: sopel/logger.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Exirel) [reply sopel/logger.py:113]: So it's only useful in debug mode, if you have some env var setup to get debug information. It's really just for development purpose, nothing comes out otherwise, as far as I can tell.

File: sopel/irc/backends.py (hunk 7 above)
Comment (Exirel) [reply sopel/irc/backends.py:277]: I agree with you. Here is what I'll do and what I suggest to do:

* in this PR, I'll keep as much as the previous behavior as possible (i.e. I'll keep the whole `ca_certs` things for now)
* in a future PR, we should rework that specific section, and everything related to that

Removing old behaviors that comes from older version of Python is, in my opinion, a good thing. It's even better to do so regarding ssl, as I know it was kind of a nightmare to deal with back in Python 2.4/2.7.

File: sopel/irc/__init__.py [hunk 8]
```diff
@@ -30,83 +30,91 @@
 import os
 import threading
 import time
-from typing import Optional
+from typing import Any, Callable, Dict, Iterable, List, Optional, Set, Tuple, TYPE_CHECKING
 
 from sopel import tools, trigger
 from sopel.tools import identifiers
```
[the remark this answers is not in the document]
Comment (dgw) [reply sopel/irc/__init__.py:36]: A change to this import from my initial review was applied in 596a58b23052f5acd2eea1253fe5268c6614758f but then reverted by 4597533fc0cde79c099923147d56a965b503e78d. @Exirel Careful when squashing this, because you probably don't want to try applying "fix 1c77c67" directly to that commit. x)

**Review by dgw — APPROVED** [review]
Bit of IRC chatter to clear up one logic-related question about CA certs, but we are probably going to remove that feature so I'm not bothered about it any more. Just one line note to save some of Exi's sanity during squash.

**Review by half-duplex — APPROVED** [review]

File: sopel/irc/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Exirel) [reply sopel/irc/__init__.py:36]: Yup, revert because another part of the file is now using a bunch of features from `sopel.tools.identifiers`.
